keps/sig-cluster-lifecycle/kubeadm/20190925-component-configs.md

```diff
@@ -0,0 +1,223 @@
+User supplied configs need to be stored in config maps and have to be manually converted by users upon upgrade, while kubeadm generated ones need neither of these things.  
+As most users would choose kubeadm to generate and manage Kubernetes core component configs itself, it is a viable proposition to treat generated and user supplied configs differently.
+
+##### Generated Kubernetes Core Component Config management
+
+Whenever kubeadm does not find component configs, in its config file or in a set of well known config maps, it will opt in for generation.  
+Generated core configs won’t be stored in config maps. Instead, they will always be regenerated in their entirety. This will allow seamless component config version upgrades when users have delegated config management responsibility to kubeadm.
+
+##### User supplied Kubernetes Core Component Config management
+
+As user changes need to be maintained, user supplied configs are stored config maps.  
+However, in those cases kubeadm is not responsible for config migration upon upgrade. If the new component version is unable to consume the user supplied config, kubeadm is going to bail out with an error message, prompting the user to perform a manual config migration beforehand. This is a behavior similar to that for addon configs.
+
+##### Allow Kustomize use over core Kubernetes component configs
+
+Kubeadm shall extend its current Kustomize support to include patching of core Kubernetes component configs. Currently, this will include only kubelet’s component config. However, in the future this might extend to the API Server, Scheduler and Controller Manager as well.  
+Most notably, this feature is not going to be applied to kube-proxy’s config by kubeadm as this is an addon component config and the responsibility for this is going to be with the corresponding addon operator.
+
+Kustomize patches will be applied on both user supplied and kubeadm generated core component configs, just before they are stored in the local files from which the config will be consumed by the components. In the case of user supplied core component configs, neither the Kustomize patches, nor the result after applying them is going to be stored on the cluster. The config maps are going to contain only the vanilla user supplied component configs.
+
+If a user has provided Kustomize patches for a config version, that is different from the current one, kubeadm is going to bail out with an error, prompting the user to either delete the old patches, or to upgrade them to the new version. This is best implemented using a kubeadm pre-flight check.
+
+UX wise, this feature is going to be reusing the existing kubeadm options that were introduced with the implementation of the kubeadm advanced configurations KEP. This allows for kustomizations to be applied to component configs on a per-node basis during init, join and upgrade just before the config is supposed to be stored on the local node file.
```

Comment:
`kubeadm advanced configurations KEP` possibly link to the KEP?

Reply:
as @fabriziopandini this means that the users have to transfer patches to each node,  but they already have to do it with the current kustomize implementation (alpha).

perhaps we can also have a way to store the patches in config maps.

Reply:
Indeed. I only sync the implementations (the static pod and component config ones will reuse the existing code and UX). With that in mind, we can enable patch storing in CMs, or even, storing already patched configs in CMs by extending the Advanced configurations KEP.